♻️ Move HPC-8205 functions into hpc-api-core

s0 opened the pull request:
Following on from #32

https://github.com/UN-OCHA/hpc_service/pull/2439 Introduces a number of general-purpose functions for reading and validating particular sets of data from the database that will likely need to be used when implementing certain functionality in the new v4 API. Rather than have them live in just hpc_service, given that they already only use the v4 models, they may as well live in this repo.

@Pl217 thoughts?

s0 committed:
🎨 Reuse Database type in context

s0 committed:
✨ Introduce getAllGoverningEntitiesForPlan

s0 committed:
✨ Introduce getAndValidateAllPlanEntities

s0 committed:
✨ Introduce getAllProjectsForPlan

s0 committed:
✨ Introduce getAllAttachments

s0 commented:
Still need to make the Errors more descriptive as requested in https://github.com/UN-OCHA/hpc_service/pull/2439#pullrequestreview-786825586

s0 committed:
📝 Improve error messages for new lib functions

s0 commented:
All errors now include IDs to help debug

Pl217 commented on src/lib/data/governingEntities.ts:
```diff
@@ -0,0 +1,107 @@
+  const gevs = await database.governingEntityVersion.find({
+    where: (builder) =>
+      builder
+        .whereIn('governingEntityId', [...new Set(ges.map((ge) => ge.id))])
```
Governing entity IDs already must be unique, due to DB primary key constraints, why put them through `Set`?
```suggestion
        .whereIn('governingEntityId', ges.map((ge) => ge.id))
```

Pl217 commented on src/lib/data/governingEntities.ts:
```diff
@@ -0,0 +1,107 @@
+}): string => {
+  const prototype = prototypes.get(governingEntity.entityPrototypeId);
+  if (!prototype) {
+    throw new Error(
+      `Missing prototype for governingEntity ${governingEntity.id}`
+    );
```
I was looking at a version before latest rebase (had it pulled locally), which had old message, without the ID information. Thanks for adding this additional error message information.

I was also thinking whether we can create a function to wrap around this pattern that we commonly use, but displaying useful error message will require that method to be really ugly, and not actually provide real benefit.

We've seen this so many times:
- Use `Map.prototype.get()`
- Check for possible `undefined` value 
- `throw new Error()` with mention of an ID

I was also thinking whether we should wrap around `Map` object and provide info like `name` when creating `Map` instance. We could then use `name` to say `Missing ${_name} for ID ${id}`, but that will also limit us in ways we want to display error messages. For example here, you search for entity prototype with `get(governingEntity.entityPrototypeId)`, but then you mention governing entity in error message, which would require our abstraction to be really messy to pull this out.

s0 replied:
Hmm yeah, this does sound messy... Having a bit of a play, It's resulted in me coming up with something like:

```ts
const getRequiredData = <Id, V, P extends string>(
  map: Map<Id, V>,
  entity: {
    [key in P]: Id;
  } & {
    id: string | number;
  },
  property: P,
  mapValueKind = 'object',
  entityKind = 'entity'
): V => {
  const val = map.get(entity[property]);
  if (!val) {
    throw new Error(
      `Missing ${mapValueKind} for ${entityKind} ${entity.id}`
    );
  }
  return val;
}
```

which can be called with something like:

```ts
const prototype = getRequiredData(prototypes, governingEntity, 'entityPrototypeId')
```

Or for the exact same error message:

```ts
const prototype = getRequiredData(prototypes, governingEntity, 'entityPrototypeId', 'prototype', 'governingEntity')
```

If we then had some sort of annotated map, we could get rid of one of those arguments in the call, but that's still a pretty messy function and arguments, which makes me feel it's not worth it to just enforce consistent error messages...

s0 replied:
One other thing I've just thought of though. Given that in most cases we're just dealing with database entities, we *could* also annotate / provide some metadata in the entities returned by the database itself. Something hidden when it comes to serializing etc... but still accessible for inspection, which can include the table name, or a general purpose way of representing the object... maybe with some `toString()` implementation that prints out the ID of the entity along with the table... similar to what we pass to `genIdentifier` when validating data upon read... hmm... I'm gonna have a bit of a think...

Pl217 commented on src/lib/data/planEntities.ts:
```diff
@@ -0,0 +1,236 @@
+      planId,
+    },
+  });
+  const planEntityIDs = new Set(planEntities.map((pe) => pe.id));
```
You don't need `Set` to make this a structure of unique plan entity IDs, since uniqueness is already enforced by DB primary key constraints.

However, `Set.prototype.has()` is nicer (and I think faster) way to search compared to `Array.prototype.indexOf()`, so we can keep it.

s0 replied:
Yeah i think that was the main motivation I had for using it! :)

Pl217 commented on src/lib/data/planEntities.ts:
```diff
@@ -0,0 +1,236 @@
+  const pevs = await database.planEntityVersion.find({
+    where: (builder) =>
+      builder
+        .whereIn('planEntityId', [...new Set(planEntities.map((pe) => pe.id))])
```
No need to map more than once, we already have this set
```suggestion
        .whereIn('planEntityId', [...planEntityIDs])
```

Pl217 commented on src/lib/data/planEntities.ts:
```diff
@@ -0,0 +1,236 @@
+  const entitiesAssociation = await database.entitiesAssociation.find({
+    where: (builder) =>
+      builder
+        .whereIn('childId', [...new Set(planEntities.map((pe) => pe.id))])
```
No need to map more than once, we already have this set
```suggestion
        .whereIn('childId', [...planEntityIDs])
```

Pl217 commented on src/lib/data/planEntities.ts:
```diff
@@ -0,0 +1,236 @@
+        .andWhere('parentType', 'governingEntity')
+        .andWhere('childType', 'planEntity'),
+  });
+  const eaByChildId = organizeObjectsByUniqueProperty(
```
By how DB tables are designed, this property isn't guaranteed to be unique. Having `entitiesAssociation` table suggests we have many-to-many relationship between governing entities and plan entites.

I have queried `entitiesAssociation` table on my local backup of production data snapshot and there are no plan entities with multiple governing entity parents. Making it possible for a plan entity to have only one governing entity parent should have been enforced in DB schema, and having `parentGoverningEntityId` column in `planEntity` table would be my way to go about this.

On the other hand, in the early days, there was maybe little knowledge about how this relationship would evolve, so they kept it flexible, and I'm wondering should we do anything about it, or just assume only a single governing entity parent.

s0 replied:
Agreed. I'm happy for this to throw an error here when we encounter situations that we don't expect (which will be the behaviour if we encounter bad data).

We could add a `UNIQUE` constraint to the combination of (`childType`, `childId`) fairly easily in a migration which will probably apply without issue, and which would protect us from causing errors at this point?

Pl217 commented on src/lib/data/planEntities.ts:
```diff
@@ -0,0 +1,236 @@
+      // Able to inspect entity details
+      entityDetails.description = planEntityVersion.value.description;
+      const supportingPlanEntityIds = planEntityVersion.value.support
+        .map((s) => s.planEntityIds)
+        .flat();
```
`Array.prototype.flatMap()` "is identical to a `map()` followed by a `flat()` of depth 1, but slightly more efficient than calling those two methods separately".

And `Array.prototype.flat()` accepts `depth` parameter, which defaults to 1 when not supplied.
```suggestion
        .flatMap((s) => s.planEntityIds);
```

s0 replied:
Nice!

Pl217 commented on src/lib/data/projects.ts:
```diff
@@ -0,0 +1,93 @@
+  const projectVersions = await database.projectVersion.find({
+    where: (builder) =>
+      builder.whereIn('id', [
+        ...new Set(pvps.map((pvp) => pvp.projectVersionId)),
```
If `projectVersionId` is a unique property of `pvps`, why put this map through `Set`? :)

You could use keys of `pvpsByProjectVersionId`
```suggestion
        ...pvpsByProjectVersionId.keys(),
```

s0 replied:
As far as I can tell, `projectVersionId` is actually not unique in `projectVersionPlan`, though it does form part of a primary key in conjunction with `planId`. I confirmed this by adding two rows with the same `projectVersionId` to a table but differing `planIds`.

That being said, logically a projectVersion should only ever be associated with a single plan, so this schema doesn't make sense as it currently stands, and `projectVersionId` isn't currently duplicated anywhere, so we could add a `UNIQUE` constraint to that column too.

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+import { isRight } from 'fp-ts/lib/Either';
```
I've noticed all other files use plural names in this directory, except this one. What a nitpick :smile:

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  }),
+]);
+
+type PlanAttachmentData = t.TypeOf<typeof ATTACHMENT_DATA>;
```
I think this reference to plan attachment in type name is misleading PlanAttachmentData
```suggestion
type AttachmentData = t.TypeOf<typeof ATTACHMENT_DATA>;
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+
+type PlanAttachmentData = t.TypeOf<typeof ATTACHMENT_DATA>;
+
+type PlanAttachmentParent =
```
Again, reference to plan attachment is misleading. All attachments are attached to plan, but not all attachments are plan attachments :smile: This type illustrates it well, but its name is not reflecting that
```suggestion
type AttachmentParent =
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+ * In addition to fetching the attachments, this also validates the contents of
+ * their value field, ensuring they are of the correct type.
+ */
+export const getAllPlanAttachments = async ({
```
This is a third and last time that I talk about how I think we should not mix attachment with plan attachment and attachment of a plan :) There are lots of references like this throughout this file, I won't comment on each
```suggestion
export const getAllAttachmentsForPlan = async ({
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  parent: PlanAttachmentParent;
+  /**
+   * If set, this particular plan attachment is associated with the given
+   * governing entity
```
Maybe expand this
```suggestion
   * governing entity. This does not necessarily mean that data available in `parent` is duplicated, because if `type` is `'planEntity'`, this field can be populated via parent of a plan entity
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  database: Database;
+  planId: PlanId;
+  types: AttachmentType[];
+  version: 'latest';
```
Why `version`? Not used

Actually, each of these newly added methods have this `version` specified in param types, but not present in params and thus not used

s0 replied:
See my comment here: https://github.com/UN-OCHA/hpc_service/pull/2439#discussion_r734571242

> I was hoping to make it clear / explicit in the function definition that it's specifically getting the latest versions of everything (and not the current / published versions), and leave room for it being expanded to current later on. Is there an alternative way you think I should do this?

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  const prototypes = await database.attachmentPrototype.find({
+    where: (builder) => builder.whereIn('type', types),
+  });
+  const attachmentPrototypes = organizeObjectsByUniqueProperty(
```
`attachmentPrototypes` is more appropriate for what is called `prototypes`
```suggestion
  const attachmentPrototypesById = organizeObjectsByUniqueProperty(
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  log: SharedLogContext;
+}): Promise<PlanAttachmentResults> => {
+  const prototypes = await database.attachmentPrototype.find({
+    where: (builder) => builder.whereIn('type', types),
```
Why not filter by `planId` as well?
```suggestion
    where: (builder) => builder.whereIn('type', types).andWhere('planId', planId)
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+  const planAttachmentVersions = await database.attachmentVersion.find({
+    where: (builder) =>
+      builder
+        .whereIn('attachmentId', [
+          ...new Set(planAttachments.map((pa) => pa.id)),
+        ])
```
No need to go through `Set`. Attachment IDs are already unique
```suggestion
        .whereIn(
          'attachmentId',
          attachments.map((a) => a.id)
        )
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+      );
+    }
+    return { type: 'planEntity', id: e.id };
+  } else {
+    throw new Error(`Invalid objectType for attachment ${attachment.id}`);
+  }
```
Could also be without `else`. We had linting rule which we removed in https://github.com/UN-OCHA/hpc-repo-tools/pull/6, but I still prefer reduced nesting :)
```suggestion
  }

  throw new Error(`Invalid objectType for attachment ${attachment.id}`);
```

Pl217 commented on src/lib/data/attachment.ts:
```diff
@@ -0,0 +1,344 @@
+    value: attachmentVersion.value,
+  };
+  const decodedValue = ATTACHMENT_DATA.decode(result);
+  if (!isRight(decodedValue)) {
```
`isLeft` doesn't get any love :smile:

s0 replied:
haha, you're right reduced nnegation would be good... I typically stick with `isRight` because "right" also means "correct" in this case (i think likely by design haha), but let's switch this out because why not, it's clear enough still.

Pl217 committed:
🎨 Remove reference to plan attachments in variable names

This is done because plan attachment could mean
"attachment of a plan" and "attachment whose parent is plan"

Pl217 committed:
🎨 Reuse `planEntityIDs` from `Set`
